File: 0138_Copy List with Random Pointer/CopyListwithRandomPointer.java
class Solution {
    public Node copyRandomList(Node head) {
        if (head == null) return null;

        HashMap<Node, Node> map = new HashMap();
        Node p = head;

        while (p != null) {
            Node cloneP = new Node(p.val);
            if (!map.containsKey(p)) {
                map.put(p, cloneP);
            }
            p = p.next;
        }

        p = head;

        while (p != null && p.next != null) {
            map.get(p).next = map.get(p.next);
            p = p.next;
        }

        p = head;

        while(p !=null){
            if(p.random!=null) map.get(p).random = map.get(p.random);
            p = p.next;
        }

        return map.get(head);
    }
}

File: 0138_Copy List with Random Pointer/CopyListwithRandomPointer.cs
public class Solution {
   public Node CopyRandomList(Node head) {
       if(head == null) return null;
       var current = head;
       while(current!=null){
           var cloneNode = new Node(current.val);
           var next = current.next;
           current.next = cloneNode;
           cloneNode.next = next;
           current = next;
       }

       current = head;
       while(current!=null){
           var next = current.next;
           if(current.random!=null){
               next.random = current.random.next;
           }           
           current = current.next.next;
       }

       current = head;
       var cloneHead = head.next;
       var cloneCurrent = cloneHead;
       while(current!=null){
           current.next = cloneCurrent.next;
           current = cloneCurrent.next;
           if(current!=null){
               cloneCurrent.next = current.next;
               cloneCurrent = cloneCurrent.next;
           }           
       }

       return cloneHead;
   }
}
